Wasmtime: Add a `gc` cargo feature

fitzgen opened the pull request:
This controls whether support for `ExternRef` and its associated deferred, reference-counting garbage collector is enabled at compile time or not. It will also be used for similarly for Wasmtime's full Wasm GC support as that gets added.

<!--
Please make sure you include the following information:

- If this work has been discussed elsewhere, please include a link to that
  conversation. If it was discussed in an issue, just mention "issue #...".

- Explain why this change is needed. If the details are in an issue already,
  this can be brief.

Our development process is documented in the Wasmtime book:
https://docs.wasmtime.dev/contributing-development-process.html

Please ensure all communication follows the code of conduct:
https://github.com/bytecodealliance/wasmtime/blob/main/CODE_OF_CONDUCT.md
-->

fitzgen commented:
I didn't add a smoke test build without the `gc` cargo feature to CI, but I can do that if we decide we can afford the CI time.

alexcrichton commented on crates/wasmtime/src/runtime/values.rs:
```diff
@@ -689,13 +702,25 @@ impl Ref {
                 Ok(TableElement::FuncRef(f.vm_func_ref(store).as_ptr()))
             }
+
+            #[cfg(feature = "gc")]
             (Ref::Extern(e), HeapType::Extern) => match e {
                 None => {
                     assert!(ty.is_nullable());
                     Ok(TableElement::ExternRef(None))
                 }
                 Some(e) => Ok(TableElement::ExternRef(Some(e.inner))),
             },
+
+            #[cfg(not(feature = "gc"))]
+            (Ref::Extern(e), HeapType::Extern) => match e {
+                None => bail!(
+                    "support for externref tables disabled at compile time \
+                     because the `gc` cargo feature was not enabled"
+                ),
+                Some(e) => match e._inner {},
+            },
```
I realize that this is what I recommended, having an uninhabited public type, but I'm second guessing myself now. This feels like it's pushing a lot of `#[cfg]` noise into the consumers of externref.

Could the methods on `ExternRef`, at least used by Wasmtime itself, be preserved? That way only `ExternRef` itself needs `#[cfg]` and consumers don't have to mind whether it's defined or not.

alexcrichton commented on crates/wasmtime/src/runtime/store.rs:
```diff
@@ -794,6 +796,9 @@ impl<T> Store<T> {
     /// automatically happen when internal buffers fill up. This is provided if
     /// fine-grained control over the GC is desired.
+    ///
+    /// This method is only available when the `gc` Cargo feature is enabled.
```
You can actually skip this documentation as rustdoc will auto-generate it with the right annotations. Elsewhere in the codebase we've got `#[cfg_attr(docsrs, doc(cfg(feature = "async")))]`, so could all public methods gated behind the `gc` feature have a similar annotation?

fitzgen replied:
It requires the `doc_cfg` nightly feature for rustdoc to add that annotation, which `docs.rs` enables but local `cargo doc` invocations probably don't. Therefore, I think it is worth leaving the comment there.

alexcrichton replied:
True yeah, but that's why it's gated by `docsrs` so the docs.rs docs are the only ones that show this. I'd recommend using the `#[cfg_attr]` approach since that's what everything else does in the API docs right now, or otherwise we should remove all those and add manual docs there too

alexcrichton commented on crates/wasmtime/src/runtime/store.rs:
```diff
@@ -1523,7 +1535,8 @@ impl StoreOpaque {
     }
 
-    pub unsafe fn insert_vmexternref_without_gc(&mut self, r: VMExternRef) {
+    #[cfg(feature = "gc")]
```
In the interest of reducing `#[cfg]` as much as possible (or at least centralizing it) could this and other helper methods be preserved here? That way in `wasmtime-runtime` itself it would have the `match x {}` to assert unreachability (or some method on `VMExternRef` itself)

fitzgen replied:
The problem with this specific case is that we store the `VMExternRefActivationsTable` directly in the store, so if it is uninhabited, then the whole `Store` becomes uninhabited.

fitzgen replied:
We could make it an `Option`, but then we enable the potential for accidentally setting it to `None` when GC *is* enabled. I think I'd rather leave these handful of `cfg`s in comparison to that.

fitzgen replied:
I guess we could have a `GcOnly<T>` type like

```rust
enum GcOnly<T> {
    #[cfg(feature = "gc")]
    inner: T,
}

impl<T> GcOnly<T> {
    pub fn with<U>(&self, f: impl FnMut(&T) -> U) -> Option<U> {
        #[cfg(feature = "gc")]
        return Some(f(&self.inner));
        #[cfg(not(feature = "gc"))]
        return None;
    }
}
```

But that seems like overkill maybe?

alexcrichton replied:
I'd personally prefer to avoid `GcOnly<T>`, I feel like those sorts of wrappers haven't worked out well in the past or require otherwise just as much `#[cfg]` traffic. For `VMExternRefActivationsTable` I'd expect that'd become an empty struct rather than uninhabited because there are no references to track?

fitzgen replied:
Ah yeah I could make that change, good idea.

fitzgen replied:
Done!

alexcrichton reviewed (commented):
Nice!

My main comments below are about perhaps shifting around `#[cfg]` to try to reduce it. That's my main worried with these sorts of features is that it can become difficult in the limit to ensure all the `#[cfg]` annotations are aligned so I'm hoping we can try to reduce them as much as possible. More-or-less I think that `ExternRef` and probably `VMExternRef` should have the same API no matter the feature set (modulo constructors) and that way consumers don't have to `#[cfg]` which should help centralize this.

We'll also definitely want this on CI and I'd recommend putting it [around here](https://github.com/bytecodealliance/wasmtime/blob/0e98a8d54ec558724481cf963f879be69ec01795/.github/workflows/main.yml#L277-L292)

fitzgen commented on crates/wasmtime/src/runtime/ref/gc_ref.rs:
```diff
@@ -0,0 +1,107 @@
+
+    /// Get the underlying data for this `ExternRef`.
+    pub fn data(&self) -> &dyn Any {
+        &*self.inner
```
I actually think we want to keep whole implementations of `ExternRef` separate for GC vs not GC. Looking into the future, there are going to be a bunch more types that are going to be doing this same split, likely whole submodules of things, and I don't think it will be very nice to have all these `cfg`s for the whole submodule tree. I think it will be better to have (a) the GC-enabled submodule tree and (b) a single GC-disabled module that defines all the same things from (a) but with uninhabited types.

fitzgen replied:
They do? The only ones left throughout the code base should be things that call constructors.

fitzgen replied:
(It just turns out we call constructors frequently, unfortunately)

fitzgen commented:
Okay, I've updated this PR to stub out more `VMExternRef` things and methods when GC is disabled at compile time and removed a bunch of `cfg`s from the rest of the code base. Not all of them in `store.rs`, as mentioned inline above. I think this is ready for another round of review.

alexcrichton commented on crates/runtime/src/table.rs:
```diff
@@ -580,7 +583,11 @@ impl Drop for Table {
+        //
+        // This is an irrefutable pattern when the `gc` cargo feature is not
+        // enabled.
+        #[allow(irrefutable_let_patterns)]
```
This is fine as-is but one trick I've done in the past is at the root of the crate:

```rust
#![cfg_attr(not(feature = "gc"), allow(irrefutable_let_patterns))]
```

or more often `unused_imports` or similar.

alexcrichton commented on crates/wasmtime/src/runtime/func.rs:
```diff
@@ -2023,6 +2027,9 @@ impl<T> Caller<'_, T> {
     ///
     /// Same as [`Store::gc`](crate::Store::gc).
+    ///
+    /// This method is only available when the `gc` cargo feature is enabled.
+    #[cfg(feature = "gc")]
```
more bits for `#[cfg_attr(docsrs, ...)]`

alexcrichton commented on crates/wasmtime/src/runtime/ref/no_gc_ref.rs:
```diff
@@ -0,0 +1,60 @@
+/// `gc` cargo feature to properly use this type.
+#[derive(Clone, Debug)]
+pub struct ExternRef {
+    pub(crate) _inner: Uninhabited,
```
Could this drop the `pub(crate)`?

alexcrichton commented on crates/wasmtime/src/runtime/ref/no_gc_ref.rs:
```diff
@@ -0,0 +1,60 @@
+
+impl ExternRef {
+    pub(crate) fn from_vm_extern_ref(_inner: VMExternRef) -> Self {
+        unreachable!()
```
Could this perhaps to `_inner.assert_unreachable()` or something like that? (where `VMExternRef` only has that method when the `gc` feature is disabled)
